### PR TITLE
Add stakingActivated Flag in `MerkleNetworkContext` for Staking

### LINKED ISSUE #3057: Add `stakingActivated` flag to state `MerkleNetworkContext`
### Problem

A reconnected node cannot assume staking is inactive just because the `0.0.800` balance is below the value of `staking.startThreshold`.

It is possible staking was previously activated, and the `0.0.800` balance has simply fallen below the start threshold.

### Solution

Add a `boolean stakingActivated` to state in the `MerkleNetworkContext`; set `stakingActivated = true` when committing the first transaction for which the `0.0.800` balance meets the value of `staking.startThreshold`.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #3057 

1. Added `stakingRewardsActivated` flag to `MerkleNetworkContext`
2. Changed `AccountsCommitInterceptor` to look at the account `0.0.800` for every balance change and checks if it reached `staking.startThreshold`. If it reaches the threshold sets `stakingActivated =true` and clears the `rewardSumHistory` array on all nodes in `MerkleStakingInfo`
3. Added logs for current testing

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/context/MutableStateChildren.java [hunk 1]
```diff
@@ -213,4 +213,9 @@ public void updatePrimitiveChildrenFrom(final ServicesState state) {
 		aliases = new WeakReference<>(state.aliases());
 		stakingInfo = new WeakReference<>(state.stakingInfo());
 	}
+
+	/* --- used only in unit tests ---*/
+	public void setNetworkCtx(final MerkleNetworkContext networkCtx) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/context/MutableStateChildren.java:218]: We can annotate with `@VisibleForTesting` to make this more visible and help the IDE find all such methods.

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java [hunk 2]
```diff
@@ -54,13 +79,32 @@ public AccountsCommitInterceptor(final SideEffectsTracker sideEffectsTracker) {
...
 		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
 			trackBalanceChangeIfAny(
 					pendingChanges.id(i).getAccountNum(),
 					pendingChanges.entity(i),
 					pendingChanges.changes(i));
 		}
 		assertZeroSum();
+		if (shouldActivateStakingRewards()) {
+			networkCtx.get().setStakingRewards(true);
+			stakingInfo.get().forEach((entityNum, info) -> info.clearRewardSumHistory());
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:95]: I see this in the HIP as well, but there was a meeting where I think Richard questioned if we want to do this. Do you remember that?

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java [hunk 3]
```diff
@@ -41,9 +49,26 @@
  */
 public class AccountsCommitInterceptor implements CommitInterceptor<AccountID, MerkleAccount, AccountProperty> {
 	private final SideEffectsTracker sideEffectsTracker;
+	private final Supplier<MerkleNetworkContext> networkCtx;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo;
+	private final GlobalDynamicProperties dynamicProperties;
+	private boolean rewardsActivated;
+	private boolean rewardBalanceChanged;
+	private long newRewardBalance;
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:57]: We could just reset this field to `-1` at the beginning of each call to `preview()`; and then simplify `shouldActivateStakingRewards()` to,
```
	public boolean shouldActivateStakingRewards() {
		return !rewardsActivated && newRewardBalance >= dynamicProperties.getStakingStartThreshold();
	}
```

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java [hunk 4]
```diff
@@ -70,6 +114,10 @@ private void trackBalanceChangeIfAny(
 	) {
 		if (accountChanges.containsKey(AccountProperty.BALANCE)) {
 			final long newBalance = (long) accountChanges.get(AccountProperty.BALANCE);
+			if (merkleAccount != null && (accountNum == STAKING_FUNDING_ACCOUNT_NUMBER)) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:117]: The staking reward account must always exist, so `null`-check is unnecessary.

File: hedera-node/src/main/java/com/hedera/services/context/MutableStateChildren.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/context/MutableStateChildren.java:218]: Fixed !

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:95]: I wasn't aware of that 👀

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:57]: Fixed !

File: hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java (hunk 4 above)
Comment (Neeharika-Sompalli) [reply hedera-node/src/main/java/com/hedera/services/ledger/interceptors/AccountsCommitInterceptor.java:117]: Fixed !

**Review by tinker-michaelj — APPROVED** [review]
LGTM 🥂

**Pull request merged**